src/endpoint/s3/ops/s3_put_bucket_lifecycle.js

```diff
@@ -8,235 +8,435 @@ const dbg = require('../../../util/debug_module')(__filename);
 const S3Error = require('../s3_errors').S3Error;
 
 const true_regex = /true/i;
+const MAX_LIFECYCLE_RULES = 1000; // AWS limit
+const MAX_TAGS_IN_AND_FILTER = 10;
 
 /**
- * validate_lifecycle_rule validates lifecycle rule structure and logical constraints
- *
- * validations:
- * - ID must be ≤ MAX_RULE_ID_LENGTH
- * - Status must be "Enabled" or "Disabled"
- * - multiple Filters must be under "And"
- * - only one Expiration field is allowed
- * - Expiration.Date must be midnight UTC format
- * - AbortIncompleteMultipartUpload cannot be combined with Tags or ObjectSize filters
+ * @param {*} field 
+ * @param {(any) => any} field_parser 
+ * @returns {any}
+ */
+function parse_lifecycle_field(field, field_parser = parseInt) {
+    if (field?.length === 1) {
+        const parsed_value = field_parser(field[0]);
+        // Basic NaN check for parseInt results
+        if (field_parser === parseInt && isNaN(parsed_value)) {
+            dbg.error(`Invalid integer value provided for field: ${field[0]}`);
+            // Use generic MalformedXML or InvalidArgument based on context
+            throw new S3Error(S3Error.InvalidArgument);
+        }
+        return parsed_value;
+    }
+    return undefined;
+}
+
+function validate_lifecycle_expiration_rule(rule, bucket_versioning) {
+    if (rule.Expiration?.length === 1) {
+        const expiration_content = rule.Expiration[0];
+        const expiration_keys = Object.keys(expiration_content);
+        if (expiration_keys.length > 1) {
+            dbg.error('Rule Expiration must specify only one of: Days, Date, or ExpiredObjectDeleteMarker', rule);
+            throw new S3Error(S3Error.MalformedXML);
+        }
+        if (expiration_content.Date) {
+            const date = new Date(expiration_content.Date[0]);
+            if (isNaN(date.getTime()) || date.getTime() !== Date.UTC(date.getUTCFullYear(), date.getUTCMonth(), date.getUTCDate())) {
+                dbg.error('Expiration Date value must conform to the ISO 8601 format and be at midnight UTC. Provided:', expiration_content.Date[0]);
+                throw new S3Error({ ...S3Error.InvalidArgument, message: "'Date' must be at midnight GMT" });
+            }
+        }
+        if (expiration_content.Days) {
```

Comment:
Does this need to be else if? also in line 55

Reply:
I don't think that would effect the correctness. We check on line 37 that there is only one key.